File: src/main/kotlin/org/libraryweasel/ligature/Ligature.kt
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package org.libraryweasel.ligature

import kotlinx.coroutines.flow.Flow

sealed class Object
data class Entity(val identifier: String): Object() {
    init {
        require(validIdentifier(identifier)) {
            "Invalid identifier: $identifier"
        }
    }
}
sealed class Literal: Object()
data class LangLiteral(val value: String, val langTag: String): Literal() {
    init {
        require(validLangTag(langTag)) {
            "Invalid lang tag: $langTag"
        }
    }
}
data class StringLiteral(val value: String): Literal()
data class BooleanLiteral(val value: Boolean): Literal()
data class LongLiteral(val value: Long): Literal()
data class DoubleLiteral(val value: Double): Literal()

data class Predicate(val identifier: String) {
    init {
        require(validIdentifier(identifier)) {
            "Invalid identifier: $identifier"
        }
    }
}

val a = Predicate("_a")
val default = Entity("_")

data class Statement(val subject: Entity, val predicate: Predicate, val `object`: Object, val context: Entity)

sealed class Range<T>(open val start: T, open val end: T)
data class LangLiteralRange(override val start: LangLiteral, override val end: LangLiteral): Range<LangLiteral>(start, end)
data class StringLiteralRange(override val start: String, override val end: String): Range<String>(start, end)
data class LongLiteralRange(override val start: Long, override val end: Long): Range<Long>(start, end)
data class DoubleLiteralRange(override val start: Double, override val end: Double): Range<Double>(start, end)

interface LigatureStore {
    fun readTx(): ReadTx
    fun writeTx(): WriteTx

    /**
     * Close connection with the Store.
     */
    fun close()
}

interface ReadTx {
    /**
     * Returns a Flow of all existing collections.
     */
    fun allCollections(): Flow<Entity>

    /**
     * Returns a collection based on the name passed.
     * Calling this function will not create a new collection, it just binds a Store and Collection name.
     */
    fun collectionReadTx(collectionName: Entity): CollectionReadTx

    /**
     * Cancels this transaction.
     */
    fun cancel()
}

interface WriteTx: ReadTx {
    /**
     * Deletes the collection of the name given and does nothing if the collection doesn't exist.
     */
    fun deleteCollection(collectionName: Entity)

    /**
     * Creates a new collection or does nothing if collection already exists.
     * Regardless the collection is returned.
     */
    fun collectionWriteTx(collectionName: Entity): CollectionWriteTx

    /**
     * Commits this transaction.
     */
    fun commit()
}

interface CollectionReadTx {
    val collectionName: Entity
    /**
     * Accepts nothing but returns a Flow of all Statements in the Collection.
     */
    fun allStatements(): Flow<Statement>

    /**
     * Is passed a pattern and returns a seq with all matching Statements.
     */
    fun matchStatements(subject: Entity? = null, predicate: Predicate? = null, `object`: Object? = null, context: Entity? = null): Flow<Statement>

    /**
     * Is passed a pattern and returns a seq with all matching Statements.
     */
    fun matchStatements(subject: Entity? = null, predicate: Predicate? = null, range: Range<*>, context: Entity? = null): Flow<Statement>
}

interface CollectionWriteTx: CollectionReadTx {
    /**
     * Returns a new, unique to this collection identifier in the form _:NUMBER"
     */
    fun newEntity(): Entity
    fun addStatement(statement: Statement)
    fun removeStatement(statement: Statement)
}

/**
 * Accepts a String representing an identifier and returns true or false depending on if it is valid.
 */
fun validIdentifier(identifier: String): Boolean {
    return "[a-zA-Z_][^\\s\\(\\)\\[\\]\\{\\}'\"`<>\\\\]*".toRegex().matches(identifier)
}

/**
 * Accepts a String representing a lang tag and returns true or false depending on if it is valid.
 */
fun validLangTag(langTag: String): Boolean {
    return "[a-zA-Z]+(-[a-zA-Z0-9]+)*".toRegex().matches(langTag)
}

File: src/main/scala/dev/ligature/Ligature.scala
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package dev.ligature

import cats.effect.Resource
import monix.eval.Task
import monix.reactive.Observable

sealed trait Object
sealed trait Node extends Object
sealed trait NamedNode extends Node
case class LocalNode(identifier: String) extends NamedNode
case class IRINode(iri: String) extends NamedNode
case class AnonymousNode(identifier: Long) extends Node()
sealed trait Literal extends Object
case class LangLiteral(value: String, langTag: String) extends Literal()
case class StringLiteral(value: String) extends Literal()
case class BooleanLiteral(value: Boolean) extends Literal()
case class LongLiteral(value: Long) extends Literal()
case class DoubleLiteral(value: Double) extends Literal()

object Ligature {
  val a: IRINode = IRINode("http://www.w3.org/1999/02/22-rdf-syntax-ns#type")

  /**
   * Accepts a String representing an identifier and returns true or false depending on if it is valid.
   */
  def validNamedNode(namedNode: NamedNode): Boolean = {
    namedNode match {
      case local: LocalNode => "[a-zA-Z_][^\\s()\\[\\]{}'\"`<>\\\\]*".r.matches(local.identifier)
      case iri: IRINode => ???
    }
  }

  /**
   * Accepts a String representing a lang tag and returns true or false depending on if it is valid.
   */
  def validLangTag(langTag: String): Boolean =
    "[a-zA-Z]+(-[a-zA-Z0-9]+)*".r.matches(langTag)
}

case class Statement(subject: Node, predicate: NamedNode, `object`: Object)
case class PersistedStatement(collection: LocalNode, statement: Statement, context: AnonymousNode)

trait Ligature {
  def session(): Resource[Task, LigatureSession]
}

trait LigatureSession {
  def read(): Resource[Task, ReadTx]
  def write(): Resource[Task, WriteTx]
}

trait ReadTx {
  /**
   * Returns a Iterable of all existing collections.
   */
  def collections(): Observable[LocalNode]

  /**
   * Returns a Iterable of all existing collections that start with the given prefix.
   */
  def collections(prefix: LocalNode): Observable[LocalNode]

  /**
   * Returns a Iterable of all existing collections that are within the given range.
   * `from` is inclusive and `to` is exclusive.
   */
  def collections(from: LocalNode, to: LocalNode): Observable[LocalNode]

  /**
   * Accepts nothing but returns a Iterable of all Statements in the Collection.
   */
  def allStatements(collection: LocalNode): Observable[PersistedStatement]

  /**
   * Is passed a pattern and returns a seq with all matching Statements.
   */
  def matchStatements(collection: LocalNode,
                      subject: Option[Node] = None,
                      predicate: Option[NamedNode] = None,
                      `object`: Option[Object] = None): Observable[PersistedStatement]

  //  /**
  //   * Is passed a pattern and returns a seq with all matching Statements.
  //   */
  //  fun matchStatements(collection: NamedEntity,
  //                      subject: Option[Entity],
  //                      predicate: Option[Predicate],
  //                      range: ClosedRange[RangeLiteral]): Any, Throwable, Observable[PersistedStatement]

  /**
   * Returns the Statement with the given context.
   * Returns None if the context doesn't exist.
   */
  def statementByContext(collection: LocalNode, context: AnonymousNode): Task[Option[PersistedStatement]]
}

trait WriteTx {
  /**
   * Creates a collection with the given name or does nothing if the collection already exists.
   * Only useful for creating an empty collection.
   */
  def createCollection(collection: LocalNode): Task[LocalNode]

  /**
   * Deletes the collection of the name given and does nothing if the collection doesn't exist.
   */
  def deleteCollection(collection: LocalNode): Task[LocalNode]

  /**
   * Returns a new, unique to this collection, AnonymousEntity
   */
  def newEntity(collection: LocalNode): Task[AnonymousNode]

  def addStatement(collection: LocalNode, statement: Statement): Task[PersistedStatement]

  //  Commenting out the below as part of #125
  //  fun removeStatement(collection: NamedEntity, statement: Statement): Any, Throwable, Statement>>
  //  fun removeEntity(collection: NamedEntity, entity: Entity): Any, Throwable, Entity>>
  //  fun removePredicate(collection: NamedEntity, predicate: Predicate): Any, Throwable, Predicate>>

  /**
   * Cancels this transaction.
   */
  def cancel(): Unit
}
